federation/pkg/kubefed/init/init.go

```diff
@@ -229,7 +230,25 @@ func initFederation(cmdOut io.Writer, config util.AdminConfig, cmd *cobra.Comman
 		return err
 	}
 
+	//to talk to federation ctrl plane
+	fedFactory := config.HostFactory(initFlags.Name, initFlags.Kubeconfig)
```

Comment:
Why build it here and not inside the `if`-block, right before the call to `waitSrvHealthy()`? I think it is Ok to just pass the context name and kubeconfig you are passing to `HostFactory()` to `waitSrvHealthy()` instead and build the client object there. 

Also, I am somewhat worried about doing this. `HostFactory()` method was written keeping interactions with host cluster in mind. It wasn't for federation API server. Also, the discovery interface for the command factory builds the objects for Kubernetes APIs and not federation APIs. All of these might work today because of the particular way things are hooked up, but might break down when some of these assumptions are violated.

In order to avoid this problem in the future and also to do the right thing, directly build the federation Clientset in `k8s.io/kubernetes/federation/client/clientset_generated/federation_clientset`. You can then call `fedClientset.Discovery()` to get the federation discovery client.